### PR TITLE
Support config.js, introduce context type for global state, and export config types

### PR DESCRIPTION
**Pull request by hasparus**
### (breaking) `--out` flag in CLI is no longer supported in favor of `config.outputPath`

We print an error message promising migration notes in [/releases](https://github.com/dethcrypto/eth-sdk/releases) so I have to write a changeset for it.

image

### configuration can now be written in JavaScript and check with `@ts-check`

Supporting `.ts` can be implemented in a separate PR.

image
 
 
 ### Introduced EthSdkCtx
 
 Big functions that know about everything anyway can stop accepting 5 args.
 
 ```ts
interface EthSdkCtx {
  readonly cliArgs: EthSdkCliArgs
  readonly config: EthSdkConfig
  readonly fs: Fs
}
```

### Updated TypeScript to v4.4 and turned on `exactOptionalPropertyTypes`

It's still a small codebase so there's no cost here, but it might catch few bugs in the future.

### Added _.vscode/tasks.json_ and _.vscode/settings.json_

- _settings.json_ states that we want to use TypeScript version from _node_modules_
- _tasks.json_ adds tasks to populate VSCode _"Problems"_ tab with ESLint and TypeScript errors. `"typecheck"` tasks is using `"type": "shell"` instead of `"type": "npm"` to make error paths start from root dir.

### CODE REVIEW

File: packages/eth-sdk/src/config/findConfigFile.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { join } from 'path'
+
+import { EthSdkCliArgs } from '../parseArgs'
+import { Fs } from '../peripherals/fs'
+
+const CONFIG_FILENAMES = ['contracts.json', 'contracts.js', 'config.js', 'config.json']
```
Comment (krzkaczor) [inline packages/eth-sdk/src/config/findConfigFile.ts:6]: remove support for `contracts.json`

File: packages/eth-sdk/src/config/traverse.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { AsyncOrSync } from 'ts-essentials'
+
+import type { NetworkSymbol } from '../abi-management/networks'
+import { unsafeEntries } from '../utils/unsafeKeys'
+import type { Address, EthSdKContracts, NestedAddresses } from './types'
+
+export async function traverseSdkDefinition(
```
Comment (krzkaczor) [inline packages/eth-sdk/src/config/traverse.ts:7]: rename -- we don't use SdkDefinition name anymore.

**Review by krzkaczor — APPROVED** [review]
Good job! 

I would suggest removing backward compatibility if this requires any additional work (like support for contracts.json). More code = more maintenance. We don't need to support any level of backward compatibility because the project is not that popular right now 😉

**Review by krzkaczor — CHANGES_REQUESTED** [review]
misclick

File: packages/eth-sdk/src/config/readConfig.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  const extension = extname(filePath)
+  const fileName = basename(filePath, extension)
+
+  try {
+    let json: unknown
+    if (['.json', '.js'].includes(extension)) {
+      json = requireJs(filePath)
+
+      // We support bare contracts dictionary for backwards compatibility.
+      if (fileName === 'contracts') {
```
[the remark this answers is not in the document]
Comment (hasparus) [reply packages/eth-sdk/src/config/readConfig.ts:17]: fwiw: we have a warning already

```
Error: Couldn't find a config file. Create one of config.js, config.json, eth-sdk.config.js, eth-sdk.config.json in /home/runner/work/eth-sdk/eth-sdk/examples/vite-react/eth-sdk
```

File: packages/eth-sdk/src/config/readConfig.test.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        kovan: {
+          mkr: parseAddress('0x9f8f72aa9304c8b593d555f12ef6589cc3a579a2'),
+        },
+      },
+      outputPath: './eth-sdk/client',
+    })
+  })
+})
+
+function mockRequire(filePath: string, result: unknown) {
```
[the remark this answers is not in the document]
Comment (krzkaczor) [reply packages/eth-sdk/src/config/readConfig.test.ts:87]: It's fine 👍

**Comment by hasparus** [comment]
Note; I've also exported `defineConfig` and started supporting `.cjs` extension.

**Review by krzkaczor — APPROVED** [review]
Great work!

**Review by krzkaczor — COMMENTED** [review]
One last thing: can you update readme getting started guide (`contracts.json` -> `config.json` etc.)

**Comment by hasparus** [comment]
> One last thing: can you update readme getting started guide (contracts.json -> config.json etc.)

We'll need to release fast, because everybody visiting the repo will see new readme.

**Pull request merged**